src/freenet/clients/http/ToadletContextImpl.java

```diff
@@ -376,15 +376,32 @@ static void sendReplyHeaders(OutputStream sockOutputStream, int replyCode, Strin
 			}
 		if(contentLength >= 0)
 			mvt.put("content-length", Long.toString(contentLength));
-		
-		String expiresTime;
+
+		boolean allowCaching; // For privacy reasons, only static
+							  // content may be cached
 		if (mTime == null) {
-			expiresTime = "Thu, 01 Jan 1970 00:00:00 GMT";
+			allowCaching = false;
+		} else {
+			allowCaching = true;
+		}
+		String expiresTime;
+		String cacheControl;
+		if (allowCaching) {
+			// use an expiry time of 30 day from now, about the frequency of Freenet releases
+			// Expires is needed for older browsers
+			expiresTime = TimeUtil.makeHTTPDate(System.currentTimeMillis() + DAYS.toMillis(30));
+			cacheControl = "public, max-age=" + String.valueOf(3600 * 24 * 30);
+			// Use contentlength as simplest possible hash function
+			// which needs no additional file access
+			mvt.put("ETag", String.valueOf(contentLength));
```

Comment:
Start time of the servlet would mean that the file would have to be re-loaded everytime freenet restarts, right? If that’s the case, then it would defeat the purpose: The CSS files would have to be re-requested in a situation where the user is most likely to open the browse page with the activelinks — that’s the page where cached CSS files help the most.

The length is not a good hash value, but we don’t need a good hash. All we need is that the value changes when the file is changed between releases. So it does not need to change on every possible file change, only on ones which are likely during development — and even if it does not change, the file should be re-requested once a month (though I’m not certain that I’m right on this). I did not use the md5 hash, because we don’t have the file content at this place in the code and I don’t want to do too invasive changes.

Another good value could be the build number of Freenet. If that isn’t known, the ETag should likely be avoided.

Please refrain from using a git hash: that would break my builds (and likely packages, too).